openedx/features/journals/api.py

```diff
@@ -0,0 +1,276 @@
+import hashlib
+from urlparse import urlsplit, urlunsplit
+import six
+
+from django.conf import settings
+from django.contrib.auth import get_user_model
+from django.core.cache import cache
+from django.core.exceptions import ObjectDoesNotExist
+from edx_rest_api_client.client import EdxRestApiClient
+from openedx.core.djangoapps.catalog.models import CatalogIntegration
+from openedx.core.djangoapps.site_configuration import helpers as configuration_helpers
+from openedx.core.lib.token_utils import JwtBuilder
+from slumber.exceptions import HttpClientError, HttpServerError
+
+
+LOGGER = logging.getLogger("edx.journals")
+JOURNALS_CACHE_TIMEOUT = 3600  # Value is in seconds
+JOURNALS_API_PATH = '/journal/api/v1/'
+JOURNAL_WORKER_USERNAME = 'journals_worker'
+User = get_user_model()
+
+
+class DiscoveryApiClient(object):
+    """
+    Class for interacting with the discovery service journals endpoint
+    """
+    def __init__(self):
+        """
+        Initialize an authenticated Discovery service API client by using the
+        provided user.
+        """
+        catalog_integration = CatalogIntegration.current()
+
+        # Client can't be used if there is no catalog integration
+        if not (catalog_integration and catalog_integration.enabled):
+            LOGGER.info("Unable to create DiscoveryApiClient because catalog integration not set up or enabled")
+            return None
+
+        try:
+            user = catalog_integration.get_service_user()
+        except ObjectDoesNotExist:
+            LOGGER.info("Unable to retrieve catalog integration service user")
+            return None
+
+        jwt = JwtBuilder(user).build_token([])
+        url = configuration_helpers.get_value('COURSE_CATALOG_API_URL', settings.COURSE_CATALOG_API_URL)
+        self.client = EdxRestApiClient(self.create_journals_url(url), jwt=jwt)
+
+    def create_journals_url(self, url):
+        """
+        rewrite the discovery url to point to journals endpoint
+        """
+        split_url = urlsplit(url)
+        override_url = urlunsplit((
+            split_url.scheme,
+            split_url.netloc,
+            JOURNALS_API_PATH,
+            split_url.query,
+            split_url.fragment,
+        ))
+        return override_url
+
+    def get_journals(self, orgs):
+        """
+        get_journals from discovery, filter on orgs is supplied
+        """
+        try:
+            if orgs:
+                response = self.client.journals.get(orgs=','.join(orgs), status='active')
+            else:
+                response = self.client.journals.get(status='active')
+            LOGGER.debug('response is type=%s', type(response))
+            return response.get('results')
+        except (HttpClientError, HttpServerError) as err:
+            LOGGER.exception(
+                'Failed to get journals from discovery-service [%s]',
+                err.content
+            )
+            return []
+
+    def get_journal_bundles(self, uuid=''):
```

Comment:
It's not 100% clear to me, but does this return different datatypes depending on whether or not a UUID is passed?

If so, can we write it to always return a list, even if that list only has a single value, for consistency?

Reply:
@tuchfarber 
Can you address this please?

Reply:
fixed, always returns a list now